cirq/ops/controlled_gate.py

```diff
@@ -26,38 +26,76 @@
+            control_qubits = []
+        self.control_qubits = control_qubits
+
+        if num_unspecified_control_qubits is None:
+            if len(control_qubits) is 0:
+                self.num_unspecified_control_qubits = 1
+            else:
+                self.num_unspecified_control_qubits = 0
+        else:
+            self.num_unspecified_control_qubits = \
+                 num_unspecified_control_qubits
+
+        # Flatten nested ControlledGates.
+        if isinstance(sub_gate, ControlledGate):
+            self.sub_gate = sub_gate.sub_gate # type: ignore
+            self.control_qubits += sub_gate.control_qubits
+            self.num_unspecified_control_qubits += \
+                 sub_gate.num_unspecified_control_qubits
+        else:
+            self.sub_gate = sub_gate
+
+        self.total_control_qubits = (len(self.control_qubits) +
+                                     self.num_unspecified_control_qubits)
 
     def num_qubits(self) -> int:
-        return self.sub_gate.num_qubits() + 1
+        return self.sub_gate.num_qubits() + self.num_unspecified_control_qubits
 
     def _decompose_(self, qubits):
         result = protocols.decompose_once_with_qubits(self.sub_gate,
-                                                      qubits[1:],
-                                                      NotImplemented)
+                                            qubits[self.total_control_qubits:],
+                                            NotImplemented)
+
         if result is NotImplemented:
             return NotImplemented
 
-        return [cop.ControlledOperation(qubits[0], op) for op in result]
+        decomposed = []
+        for op in result:
+            controlled_op = op
+            for qubit in qubits[:self.total_control_qubits]:
+                controlled_op = cop.ControlledOperation(qubit, controlled_op)
+            decomposed.append(controlled_op)
+        return decomposed
 
     def validate_args(self, qubits) -> None:
-        if len(qubits) < 1:
+        if len(qubits) < self.num_unspecified_control_qubits:
             raise ValueError('No control qubit specified.')
-        self.sub_gate.validate_args(qubits[1:])
+        self.sub_gate.validate_args(
+                          qubits[self.num_unspecified_control_qubits:])
 
     def _value_equality_values_(self):
-        return self.sub_gate
+        return (self.sub_gate, self.total_control_qubits)
 
     def _apply_unitary_(self, args: protocols.ApplyUnitaryArgs) -> np.ndarray:
-        qubits = cirq.LineQubit.range(1 + self.sub_gate.num_qubits())
-        c_op = cop.ControlledOperation(qubits[0],
-                                       self.sub_gate.on(*qubits[1:]))
+        qubits = cirq.LineQubit.range(self.total_control_qubits +
+                                      self.sub_gate.num_qubits())
+        c_op = self.sub_gate.on(*qubits[self.total_control_qubits:])
+        for control in qubits[:self.total_control_qubits]:
```

Comment:
This loop can also be avoided if GateOperation is made able to handle multiple controls.

Reply:
Agreed. Will change after that is implemented.